add foreach, foreach_parallel and foreach_parallel_n_threads (fix #260)

pthom opened the pull request:
See https://github.com/Dobiasd/FunctionalPlus/issues/260

Example usage:

````cpp
 constexpr int rows = 3000, cols = 2000;
 MandelbrotMat<rows, cols> mandelbrot_mat;
 auto fill_mandelbrot_row = [& mandelbrot_mat ](int row_idx) {
        ...
 };
 mandelbrot_mat = MandelbrotMat<rows, cols>();
 auto rows_idxs = fplus::numbers(0, rows);
 fplus::parallel_for(fill_mandelbrot_row, rows_idxs);
````

Dobiasd commented on include/fplus/side_effects.hpp:
```diff
@@ -283,6 +283,46 @@ auto execute_n_times(std::size_t n, const Effect& eff)
+// auto rows_idxs = fplus::numbers(0, rows);
+// fplus::parallel_for(fill_mandelbrot_row, rows_idxs);
+template<typename F, typename ContainerIdxs>
+void parallel_for(F f, const ContainerIdxs& idxs_to_iterate)
```
I wouldn't even call it `ContainerIdxs& idxs_to_iterate`, but just `Container& xs`. Of course, the elements can be indexes, but they don't have to. :)

Dobiasd commented on test/side_effects_test.cpp:
```diff
@@ -76,3 +76,49 @@ TEST_CASE("side_effects_test - execute_parallelly_atomic")
+{
+    using namespace fplus;
+
+    constexpr int rows = 3000, cols = 2000;
```
If I understand correctly, on a 64-bit system, this will use 48 MB. How about a smaller, maybe even generally simpler, test?

Dobiasd commented on include/fplus/side_effects.hpp:
```diff
@@ -283,6 +283,46 @@ auto execute_n_times(std::size_t n, const Effect& eff)
+// using `n_threads` threads.
+// The function will perform its side effects, and nothing is returned.
+template<typename F, typename ContainerIdxs>
+void parallel_for_n_threads(size_t n_threads, F f, const ContainerIdxs& idxs_to_iterate)
```
What do you think about naming there functions
- `parallel_for_each`
- `parallel_for_each_n_threads`

instead?

It expresses the "execute for each element in this generic container", and also seems to be kind of a default name for such a function.

We could even introduce a normal (non-parallel) `fplus::for_each` too, so that it would align with the existing three `fplus::transform ...` functions. :slightly_smiling_face:

pthom commented:
Yes to all your comments !

So, I changed the names to for_each, for_each_parallel and for_each_parallel_n_threads.

I removed the example from the API doc, as I 'm not sure it was that interesting.

Dobiasd commented:
Thanks, looks good to me now. :+1:

(Since the example is now in the commit message, it won't be lost completely. ^^)

The original test was also interesting (at least to me), but the new one feels more standard, simpler, and boring. Boring code is good code. :grin:
